Utilities/MachineInventory.cs

```diff
@@ -0,0 +1,245 @@
+        }
+
+        /// <summary>
+        ///     Remove items from inventory if items are not on the blacklist
+        /// </summary>
+        /// <param name="items">The target inventory or list to store the items</param>
+        /// <param name="blacklist">The list of items forbidden from transfer</param>
+        /// <param name="storagecapacity">Storage capacity of target inventory</param>
+        /// <param name="amount">Amount of items to move in this transfer</param>
+        public void RemoveBlackList(IEnumerable<ItemBase> items, IEnumerable<ItemBase> blacklist, int storagecapacity = int.MaxValue, int amount = 1)
+        {
+            ItemBaseUtil.MoveItems(this.Inventory, items, amount, storagecapacity, false, blacklist, false);
+        }
+
+        /// <summary>
+        ///     Remove items from inventory if items are not on the blacklist
+        /// </summary>
+        /// <param name="items">The target inventory or list to store the items</param>
+        /// <param name="blacklist">Item forbidden from transfer</param>
+        /// <param name="storagecapacity">Storage capacity of target inventory</param>
+        /// <param name="amount">Amount of items to move in this transfer</param>
+        public void RemoveBlackList(IEnumerable<ItemBase> items, ItemBase blacklist, int storagecapacity = int.MaxValue, int amount = 1)
+        {
+            ItemBase[] BlackList = new[] { blacklist };
+            ItemBaseUtil.MoveItems(this.Inventory, items, amount, storagecapacity, false, BlackList, false);
+        }
+
+        /// <summary>
+        ///     Returns the spare capacity of the inventory
+        /// </summary>
+        /// <returns>The spare capacity of the inventory</returns>
+        public int SpareCapacity()
+        {
+            return this.StorageCapacity - this.Inventory.GetItemCount();
+        }
+
+        /// <summary>
+        ///     Returns the current number of items in the inventory
+        /// </summary>
+        /// <returns>Current number of items in inventory</returns>
+        public int ItemCount()
+        {
+            return this.Inventory.GetItemCount();
+        }
+
+        public void WriteInventory(BinaryWriter writer)
+        {
+            int listcount = this.Inventory.Count;
+            int version = 0;
+
+            writer.Write(version);
+            writer.Write(listcount);
+            for (int index = 0; index < listcount; ++index)
+            {
+                ItemFile.SerialiseItem(this.Inventory[index], writer);
+            }
+        }
+
+        public void ReadInventory(BinaryReader reader)
+        {
+            int listcount;
+            int version = reader.ReadInt32();
+
+            switch (version)
+            {
+                case 0:
+                    listcount = reader.ReadInt32();
+                    for (int index = 0; index < listcount; ++index)
+                    this.Inventory.Add(ItemFile.DeserialiseItem(reader));
```

Comment:
If the game corrupts for whatever reason, and the thing it reads is invalid, the ItemFile.DeserializeItem returns null. I suggest having:

var item = ItemFile.DeserializeItem(reader);
if(item != null) this.Inventory.Add(item);